editor/inspector/editor_preview_plugins.cpp

```diff
@@ -292,28 +321,1071 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
 }
 
 Ref<Texture2D> EditorPackedScenePreviewPlugin::generate_from_path(const String &p_path, const Size2 &p_size, Dictionary &p_metadata) const {
-	String temp_path = EditorPaths::get_singleton()->get_cache_dir();
-	String cache_base = ProjectSettings::get_singleton()->globalize_path(p_path).md5_text();
-	cache_base = temp_path.path_join("resthumb-" + cache_base);
+	ERR_FAIL_COND_V_MSG(!Engine::get_singleton()->is_editor_hint(), Ref<Texture2D>(), "This function can only be called from the editor.");
+	ERR_FAIL_NULL_V_MSG(EditorNode::get_singleton(), Ref<Texture2D>(), "EditorNode doesn't exist.");
 
-	//does not have it, try to load a cached thumbnail
+	// Lower the abort flag only if it's not raised (beware the mutex, don't use `_get_aborted()`)
+	abort_mutex->lock();
+	if (aborted) {
+		abort_mutex->unlock();
+		return Ref<Texture2D>();
+	}
+	aborted = false;
+	abort_mutex->unlock();
 
-	String path = cache_base + ".png";
+	// Estimate scene complexity before loading it
+	uint64_t file_size_acc = _get_accumulated_file_size(p_path);
 
-	if (!FileAccess::exists(path)) {
+	// Do not create thumbnail if scene file size exceeds threshold (use dummy to avoid retry loop)
+	if (file_size_acc > uint64_t(EDITOR_GET("docks/filesystem/thumbnail_file_size_threshold"))) {
+		return _create_dummy_thumbnail();
+	}
+
+	if (_get_aborted()) {
+		print_line("Thumbnail creation aborted.");
```

Comment:
I think these all might be better served as verbose prints
```suggestion
		print_verbose("Thumbnail creation aborted.");
```
If not done as error or warnings